rig-lancedb/src/lib.rs

```diff
@@ -0,0 +1,300 @@
+    /// See [LanceDb pre/post filtering](https://lancedb.github.io/lancedb/sql/#pre-and-post-filtering) for more information
+    post_filter: Option<bool>,
+    column: Option<String>,
+}
+
+impl SearchParams {
+    pub fn distance_type(mut self, distance_type: DistanceType) -> Self {
+        self.distance_type = Some(distance_type);
+        self
+    }
+
+    pub fn search_type(mut self, search_type: SearchType) -> Self {
+        self.search_type = Some(search_type);
+        self
+    }
+
+    pub fn nprobes(mut self, nprobes: usize) -> Self {
+        self.nprobes = Some(nprobes);
+        self
+    }
+
+    pub fn refine_factor(mut self, refine_factor: u32) -> Self {
+        self.refine_factor = Some(refine_factor);
+        self
+    }
+
+    pub fn post_filter(mut self, post_filter: bool) -> Self {
+        self.post_filter = Some(post_filter);
+        self
+    }
+
+    pub fn column(mut self, column: &str) -> Self {
+        self.column = Some(column.to_string());
+        self
+    }
+}
+
+impl<M: EmbeddingModel> LanceDbVectorStore<M> {
+    pub async fn new(
+        table: lancedb::Table,
+        model: M,
+        id_field: &str,
+        search_params: SearchParams,
+    ) -> Result<Self, lancedb::Error> {
+        Ok(Self {
+            table,
+            model,
+            id_field: id_field.to_string(),
+            search_params,
+        })
+    }
+
+    /// Define an index on the specified fields of the lanceDB table for search optimization.
+    /// Note: it is required to add an index on the column containing the embeddings when performing an ANN type vector search.
+    pub async fn create_index(
+        &self,
+        index: Index,
+        field_names: &[impl AsRef<str>],
+    ) -> Result<(), lancedb::Error> {
+        self.table.create_index(field_names, index).execute().await
+    }
+}
+
+impl<M: EmbeddingModel + std::marker::Sync + Send> VectorStoreIndex for LanceDbVectorStore<M> {
+    async fn top_n<T: for<'a> Deserialize<'a> + std::marker::Send>(
+        &self,
+        query: &str,
+        n: usize,
+    ) -> Result<Vec<(f64, String, T)>, VectorStoreError> {
+        let prompt_embedding = self.model.embed_document(query).await?;
+
+        let query = self
+            .table
+            .vector_search(prompt_embedding.vec.clone())
+            .map_err(lancedb_to_rig_error)?
+            .limit(n);
+
+        self.build_query(query)
+            .execute_query()
+            .await?
+            .into_iter()
+            .map(|value| {
+                Ok((
+                    match value.get("_distance") {
+                        Some(Value::Number(distance)) => distance.as_f64().unwrap_or_default(),
+                        _ => 0.0,
+                    },
+                    match value.get(self.id_field.clone()) {
+                        Some(Value::String(id)) => id.to_string(),
+                        _ => "".to_string(),
+                    },
```

Comment:
Duplicate ids will cause an error for certain providers (e.g.: cohere) so the default returned here should not be the empty string, but something unique.

Suggestion: call enumerate before this map and use the value `unknown{n}` as the default id value